### PR TITLE
recursive feature elimination for feature selection closes #134

### LINKED ISSUE #134: feature selection: recursive feature elimination
A transformer that implements this for various algorithms:

https://github.com/solegalli/feature-selection-for-machine-learning/blob/master/11-Hybrid-methods/11.2-Recursive-feature-elimination.ipynb

Should be able to use ML algos that output some sort of importance value, like coef_ or feature_importance_

### PR DESCRIPTION
**Pull request by SunnyxBd**
…ination

### CODE REVIEW

File: feature_engine/selection/eliminate_feature_by_feature_importance.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        )
+
+        # store initial model performance
+        self.initial_model_performance_ = model["test_score"].mean()
+
+        # get performance metric
+        scorer = get_scorer(self.scoring)
+
+        # choose the approriate attibute to use for the appropriate model
+        if (self.regression):
```
Comment (solegalli) [inline feature_engine/selection/eliminate_feature_by_feature_importance.py:135]: we can get the attributes from the estimators directly like this:
https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/feature_selection/_from_model.py#L18

**Comment by SunnyxBd** [comment]
@solegalli 

Here's my first version of the recursive feature elimination selector.

The code is functional but lacking explicative comments.
I intend to spend more time of bettering it tomorrow, before I start adding the unit tests.

Feel free to leave any comments

**Review by solegalli — COMMENTED** [review]
Hey @SunnyxBd this looks really good, thank you!

I added a few suggestions. Let me know your thoughts. Cheers

File: feature_engine/selection/eliminate_feature_by_feature_importance.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+            feature_importance_score = {k: tmp.get(
+                k, 0) + feature_importance_score.get(k, 0) for k in set(tmp)}
+
+        self.ordered_features_by_importance_ = [k for k, v in sorted(
+            feature_importance_score.items(), key=lambda item: item[1])]
+
+        # list to collect selected features
+        self.selected_features_ = []
+
+        X_amputated = X.copy()
```
[the remark this answers is not in the document]
Comment (SunnyxBd) [reply feature_engine/selection/eliminate_feature_by_feature_importance.py:161]: hahaha! Sure thing!

**Comment by SunnyxBd** [comment]
@solegalli 

Thank you a lot for your comments, I'm going to go through them one by one 😸

File: feature_engine/selection/eliminate_feature_by_feature_importance.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            features = pd.DataFrame(data=getattr(
+                m, get_feature_importance_method), columns=["coef_importance"])
+            features.index = self.variables
+            features.sort_values(ascending=True, inplace=True, by='coef_importance')
+
+            features['variable_score'] = np.arange(features.shape[0])
+
+            tmp = features['variable_score'].to_dict()
+            feature_importance_score = {k: tmp.get(
+                k, 0) + feature_importance_score.get(k, 0) for k in set(tmp)}
```
[the remark this answers is not in the document]
Comment (SunnyxBd) [reply feature_engine/selection/eliminate_feature_by_feature_importance.py:153]: I like this approach better, thanks!

Since i have created in my first version an attribute that returns the ordered features by importance list, called "feature_importances_", do you we should get rid of it if we are also returning the whole ordered series with features and feature importance values?

It might be redundant if we keep both. What do you think ?

File: feature_engine/selection/eliminate_feature_by_feature_importance.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        self.initial_model_performance_ = model["test_score"].mean()
+
+        # get performance metric
+        scorer = get_scorer(self.scoring)
+
+        # choose the approriate attibute to use for the appropriate model
+        if (self.regression):
+            get_feature_importance_method = "coef_"
+        else:
+            get_feature_importance_method = "feature_importances_"
```
[the remark this answers is not in the document]
Comment (SunnyxBd) [reply feature_engine/selection/eliminate_feature_by_feature_importance.py:133]: Also, coef_ or importances_ depend on whether it is a tree or a linear model, not on regression vs classification

> Oh, yes Definitely..

I For the coeff importance part. I have a clear idea about how to implement it now.
I'm thinking of adding a new file in the selection folder called "selection_utils"? where I would put the code for the function that returns the feature importance giving a fitted estimator.
Or should I add the function in the __init__.py file in the selection folder ?
---------------------------------------------------------------------------
Where do you see this file existing?

[the remark this answers is not in the document]
Comment (solegalli) [reply feature_engine/selection/eliminate_feature_by_feature_importance.py]: for now I would say, add a file called base_selector.py within the selection folder. The classes and functions there should not be called from the init because they are internal.

Then, as more selectors come up, we see if we need to move around. Thank you!

File: feature_engine/selection/eliminate_feature_by_feature_importance.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (solegalli) [reply feature_engine/selection/eliminate_feature_by_feature_importance.py]: If it has the same information, we should also keep one of the 2.

**Comment by solegalli** [comment]
> 
> 
> @solegalli
> 
> Thank you a lot for your comments, I'm going to go through them one by one smile_cat

Great, thank you! Let me know when I am good to review again. Cheers.

**Comment by SunnyxBd** [comment]
@solegalli 

I made progress, can you review please when you find time?

I'm going to be working on unit tests next.

File: feature_engine/selection/__init__.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 """
 The module selection includes classes to select features or remove unwanted features.
 """
-
+from .eliminate_feature_by_feature_importance import RecursiveFeatureElimination
```
Comment (solegalli) [inline feature_engine/selection/__init__.py:4]: Could we change the name of the .py to recursive_feature_elimination.py ?

The reason is that we (will) have several transformers that select features by importance, and it might be confusing to understand which one we will find in this script if we keep the generic name.

Thank you!

File: feature_engine/selection/base_selector.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+def get_feature_importances(estimator):
+
+    """Retrieve feature importances from a fitted estimator"""
+
+    importances = getattr(estimator, "feature_importances_", None)
+
+    coef_ = getattr(estimator, "coef_", None)
+
+    if importances is None and coef_ is not None:
```
Comment (solegalli) [inline feature_engine/selection/base_selector.py:12]: Question: could it not be simplified to `if coef_` ?

I think `if coef_is not None`, is captured by `if coef_`, and I don't think there is a scenario where` importances is not None` and `Coef_` is also `not None`, so not sure we need both? I might be wrong.

File: feature_engine/selection/eliminate_feature_by_feature_importance.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import cross_validate
+from sklearn.utils.validation import check_is_fitted
+
+from feature_engine.selection.base_selector import get_feature_importances
```
Comment (solegalli) [inline feature_engine/selection/eliminate_feature_by_feature_importance.py:7]: is it possible to order feature_engine imports alphabetically?

File: feature_engine/selection/eliminate_feature_by_feature_importance.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                    return_estimator=True,
+                    scoring=self.scoring
+                )
+
+                # store initial model performance
+                baseline_model_performance = baseline_model["test_score"].mean()
+
+        self.input_shape_ = X.shape
+
+        return self
```
Comment (solegalli) [inline feature_engine/selection/eliminate_feature_by_feature_importance.py:230]: excellent!!! This looks amazing, thank you so much!

File: tests/test_selection/test_eliminate_feature_by_feature_importance.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
+import numpy as np
```
Comment (solegalli) [inline tests/test_selection/test_eliminate_feature_by_feature_importance.py:1]: can we change the name of the test to test_recursive_feature_elimination.py? so that it mirrors both the name of the class and the python script if you renamed it?

File: tests/test_selection/test_eliminate_feature_by_feature_importance.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+
+@pytest.fixture(scope="module")
+def load_diabetes_dataset():
+    # Load the diabetes dataset from sklearn
+    diabetes_X, diabetes_y = load_diabetes(return_X_y=True)
+    X = pd.DataFrame(diabetes_X)
+    y = pd.DataFrame(diabetes_y)
+    return X, y
+
```
Comment (solegalli) [inline tests/test_selection/test_eliminate_feature_by_feature_importance.py:39]: if we use the same dataframes for more than 1 transformer, we should see if we can add a conf file inside the test_selection folder.

Not sure at the moment this is best practice, do you? otherwise I could do some research later today, or Monday.

File: tests/test_selection/test_eliminate_feature_by_feature_importance.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        "var_10",
+        "var_11",
+    ]
+    assert sel.threshold == 0.001
+    assert sel.cv == 3
+    assert sel.scoring == "roc_auc"
+    # test fit attrs
+    assert np.round(sel.initial_model_performance_, 3) == 0.997
+    assert sel.selected_features_ == ['var_0', 'var_6']
+    # test transform output
```
Comment (solegalli) [inline tests/test_selection/test_eliminate_feature_by_feature_importance.py:71]: this is great. If we wanted to be super thorough, maybe we should also check the attribute ordered_features_by_importance?

We didn't add this in shuffle features because with shuffling, the values change every time we run the transformer. But with this transformer, they shouldn't because we are setting the seed in the random_forest.

**Review by solegalli — CHANGES_REQUESTED** [review]
@SunnyxBd great work! These look amazing! The tests are great, and the transformer looks really good, super tidy. Thank you so much.

It is in a good place to merge. I just made a few suggestions, here and there.

We need to add these to the docs:

- [ ] docs/index.rst
- [ ] docs/selection/index.rst
- [ ] docs/selection/new_rst)with_example.rst

Would you be able to add those? 

Sorry for the delay in the review. I had a tight deadlines. And thank you so much for the amount of work

**Comment by SunnyxBd** [comment]
Thank you so much for your comments and help @solegalli .

Yes I can definitely work on those changes.

I'll start Monday if that's alright with you. 

Have a nice week-end!✨

**Comment by SunnyxBd** [comment]
@solegalli 
I've worked on the requested changes.

**if coef_** doesn't work because coef_ is an array (when it is not None) and that (if array) generates an error.
So I kept it to  **if coef_ is Not None**.

Please review my changes when you find time.

I'm going to start working on updating the docs.

Cheers,
Sana

**Comment by SunnyxBd** [comment]
@solegalli 

I finished writing the unit tests and updating the docs. Please have a look when you find time.

ci/circleci: test_docs had failed, I'm not sure why. Could you please take a look at it?

Thank you,
Sana
